File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer,Integer> map = new HashMap<>();
        PriorityQueue<Map.Entry<Integer,Integer>> pq = new PriorityQueue<>
            ((a,b) -> b.getValue() - a.getValue());
        
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        
        for (Map.Entry<Integer,Integer> entry : map.entrySet()) {
            pq.add(entry);
        }
        
        
        int[] output = new int[k];
        for (int i = 0; i < k; i++) {
            output[i] = pq.poll().getKey();
        }
        
        return output;
    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # num to frequency
        hashmap = {}
        
        # frequency to num
        freq = [[] for i in range(len(nums) + 1)]
        
        for num in nums:
            hashmap[num] = hashmap.get(num, 0) + 1
        
        for num, count in hashmap.items():
            freq[count].append(num)
            
            
        res = []
        
        for i in range(len(freq) - 1, 0, -1):
            for val in freq[i]:
                res.append(val)
                if len(res) == k:
                    return res